### PR TITLE
Optimize merge of uniqExact without_key

### LINKED ISSUE #4510: uniqExact(distinct count) is slow
I'm comparing ClickHouse with a commercial columnar database. It's amazing that in some cases ClickHouse beats the commercial product even with less resources(4 cores vs. 20 cores in my case). However, I noticed uniqExact(or distinct count) is unreasonably slow.

Here is my findings and hopefully they can be of use:
* assume we have a table with a few hundred millions rows(mine: ~600,000,000 rows) like below
```sql
create table test_distinct_count(a String, b String, b Int32, d String, ...)
Engine=MergeTree() order by (a, b, c, d) settings index_granularity = 8192;
```

* we'll likely get the following test results on a VM with 4 cores(max_threads=4)
```sql
select a from test_distinct_count group by a -- 1,100+ rows returned in 5.252s
select distinct a from test_distinct_count -- 1,100+ rows returned in 5.912s
select uniq(a) from test_distinct_count -- 1 row returned in 7.483s
select uniqExact(a) from test_distinct_count -- 1 row returned in 12.308s
select count(1) from (select distinct a from test_distinct_count) x -- 1 row returned in 5.821s
select count(1) from (select a from test_distinct_count group by a) x -- 1 row returned in 4.937s
```

It looks like using `group by` and avoiding `distinct` bring us better performance, but it's really inconvenient to replace all `distinct count` with the trick...

### PR DESCRIPTION
**Pull request by nickitat**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Parallelized merging of `uniqExact` states for aggregation without a key, i.e. queries like `SELECT uniqExact(number) FROM table`. The improvement becomes noticeable when the number of unique keys approaches 10^6.
Also `uniq` performance is slightly optimized.
This closes #4510.

---

x86: 

11.880 | 2.753 | -4.315x | -0.769 | 0.768 | uniq_without_key | 26 | SELECT uniqExact(number) from numbers_mt(1e8)
-- | -- | -- | -- | -- | -- | -- | --

arm:

5.777 | 1.077 | -5.36x | -0.814 | 0.813 | uniq_without_key | 26 | SELECT uniqExact(number) from numbers_mt(1e8)
-- | -- | -- | -- | -- | -- | -- | --

### CODE REVIEW

**Comment by nickitat** [comment]
AST fuzzer (asan) — #43199

**Comment by nickitat** [comment]
AST fuzzer (debug) - https://github.com/ClickHouse/ClickHouse/issues/42819
AST fuzzer (ubsan) - https://github.com/ClickHouse/ClickHouse/issues/42691
Integration tests (tsan) [1/4] - https://github.com/ClickHouse/ClickHouse/issues/43169
Stateless tests (release, s3 storage) - `01079_parallel_alter_detach_table_zookeeper` doesn't look relevant

File: src/AggregateFunctions/UniqExactSet.h [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+                    while (true)
+                    {
+                        const auto bucket = next_bucket_to_merge->fetch_add(1);
+                        if (bucket >= rhs.NUM_BUCKETS)
+                            return;
+                        lhs.impls[bucket].merge(rhs.impls[bucket]);
+                    }
+                };
+
+                for (size_t i = 0; i < thread_pool->getMaxThreads(); ++i)
```
Comment (KochetovNicolai) [inline src/AggregateFunctions/UniqExactSet.h:67]: This can be a dangerous code a bit. If somebody common background I/O pool with 1000 threads, it will add 1000  tasks. let's add at least <= NUM_BUCKETS tasks.

Also, with current implementation we can't share such a pool with a multiple uniq functions.
Maybe it's better to avoid calling wait() at all and handle exceptions in every task separately. (But out ThreadPool interface is not so good for it).

File: src/AggregateFunctions/AggregateFunctionUniq.h [hunk 2]
```diff
@@ -162,17 +173,29 @@ template <typename T> struct AggregateFunctionUniqTraits
 };
 
 
-/** The structure for the delegation work to add one element to the `uniq` aggregate functions.
+/** The structure for the delegation work to add elements to the `uniq` aggregate functions.
   * Used for partial specialization to add strings.
   */
-template <typename T, typename Data>
-struct OneAdder
+template <typename T, typename Data, bool is_variadic = false, bool is_exact = false, bool argument_is_tuple = false>
```
Comment (KochetovNicolai) [inline src/AggregateFunctions/AggregateFunctionUniq.h:179]: I had an idea that probably we can put this 3 flags into `Data`.
Probably it will make a code a bit more readable. (If it is possible to do).

And maybe we can do it with `is_able_to_parallelize_merge` flag as well to those Data which support it, like
```
if (settings->max_threads > 1)
            return createAggregateFunctionUniq<
                ..., AggregateFunctionUniqExactData<..., true /* is_able_to_parallelize_merge>
else
            return createAggregateFunctionUniq<
                ..., AggregateFunctionUniqExactData<..., false /* is_able_to_parallelize_merge>
```

File: src/Common/HashTable/HashTable.h [hunk 3]
```diff
@@ -1263,30 +1251,6 @@ class HashTable :
                 ptr->write(wb);
     }
 
-    void writeText(DB::WriteBuffer & wb) const
```
Comment (KochetovNicolai) [inline src/Common/HashTable/HashTable.h:1266]: I am confused a bit why this code is removed.
Probably it's not used, but may be still helpful for debugging.
I would prefer to do it in a separate pr if possible.

**Review by KochetovNicolai — APPROVED** [review]
Generally looks ok.

File: src/AggregateFunctions/UniqExactSet.h (hunk 1 above)
Comment (nickitat) [reply src/AggregateFunctions/UniqExactSet.h]: > let's add at least <= NUM_BUCKETS tasks.

ok

> Also, with current implementation we can't share such a pool with a multiple uniq functions.

yes, we cannot merge states of two different `uniqExact`-s simultaneously, but it shouldn't be a problem since we aim to utilise all the threads by the current merge. if we would share, client code would need to manually call `wait` before destroying states which is also not ideal.

File: src/AggregateFunctions/AggregateFunctionUniq.h (hunk 2 above)
Comment (nickitat) [reply src/AggregateFunctions/AggregateFunctionUniq.h]: also thought about that, looks more accurate

File: src/Common/HashTable/HashTable.h (hunk 3 above)
Comment (nickitat) [reply src/Common/HashTable/HashTable.h]: yeah, let's left it untouched

**Review by KochetovNicolai — APPROVED** [review]
